File: AppQuiz_10130492_20221207/app/src/main/java/com/example/appquiz_10130492_20221207/MainActivity_10130492.java
package com.example.appquiz_10130492_20221207;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.toolbox.JsonArrayRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

public class MainActivity_10130492 extends AppCompatActivity {

    ArrayList<BookStore> bookStores = new ArrayList<>();
    BookAdapter adapter;
    TextView loading;
    FloatingActionButton fab;
    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_10130492);
        fab = findViewById(R.id.fab);
        loading = findViewById(R.id.load);


        RecyclerView recyclerView = findViewById(R.id.rec);
        recyclerView.setVisibility(View.INVISIBLE);
        adapter = new BookAdapter(bookStores);
        loading.setVisibility(View.VISIBLE);
        getData();

        recyclerView.setVisibility(View.VISIBLE);

        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
        fab.setOnClickListener(i->{
            loading.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.INVISIBLE);
            getData();
            recyclerView.setAdapter(adapter);
            recyclerView.setVisibility(View.VISIBLE);


        });


    }
    private void getData(){
        Response.Listener<JSONArray> response_listener = response -> {
            Log.d("JSON",response.toString());
            parseJson(response);

        };
        Response.ErrorListener response_errListener = err->{
          Log.e("JSON",err.toString());
        };
        String url = "https://cloud.culture.tw/frontsite/trans/emapOpenDataAction.do?method=exportEmapJson&typeId=M";
        JsonArrayRequest jsonArrayRequest = new JsonArrayRequest(url, response_listener, response_errListener);
        RequestQueue requestQueue = Volley.newRequestQueue(this);
        requestQueue.add(jsonArrayRequest);

    }

    private void parseJson(JSONArray jsonArray){
        try{
            bookStores.clear();
            adapter.notifyDataSetChanged();
            Log.i("JSONParse",jsonArray.length()+"");
            for(int i = 0 ; i < jsonArray.length() ; i++){
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                String name = jsonObject.getString("name");
                String cityName = jsonObject.getString("cityName");
                String openTime = jsonObject.getString("openTime");
                BookStore bookStore = new BookStore(name,cityName,openTime);
                Log.d("JSONData", String.format("Name: %s, City: %s",bookStore.getName(),bookStore.getCityName()));
                bookStores.add(bookStore);

            }
            adapter.notifyDataSetChanged();
            loading.setVisibility(View.INVISIBLE);

        }catch(Exception e){

            Log.e("JSON",e.toString());
        }
    }

}

File: AppQuiz_10130492_20221228_p1/app/src/main/java/com/example/appquiz_10130492_20221228/MainActivity_10130492.kt
package com.example.appquiz_10130492_20221228

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.volley.Response
import com.android.volley.toolbox.JsonArrayRequest
import com.android.volley.toolbox.Volley
import org.json.JSONArray

class MainActivity_10130492 : AppCompatActivity() {
    private lateinit var readBtn: Button
    private lateinit var name: TextView
    private lateinit var intro: TextView
    var scenaryList = ArrayList<Scenary>()
    private val url = "https://data.coa.gov.tw/Service/OpenData/ODwsv/ODwsvAttractions.aspx"
    override fun onCreate(savedInstanceState: Bundle?) {
        context = this
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_10130492)
        readBtn = findViewById(R.id.btnRead)
        name  = findViewById(R.id.name)
        intro = findViewById(R.id.intro)
        readBtn.setOnClickListener{

            //getData_JSONObject()
            load()


        }

    }

    private fun parseJson(json: JSONArray){

        Log.i("JSON",json.toString())
        scenaryList.clear()
        for(i in 0 until json.length()){
            val jsonObj = json.getJSONObject(i)
            val s = Scenary(jsonObj.getString("Name"),jsonObj.getString("City"),jsonObj.getString("Tel"),jsonObj.getString("Address"),jsonObj.getString("Introduction"))
            name.text = s.name
            intro.text = s.description
            scenaryList.add(s)
        }



    }
    fun load(){
        val responseListener = Response.Listener<JSONArray> {
            parseJson(it)
        }
        val reponseError = Response.ErrorListener {
            it.printStackTrace()
        }
        val jsonArrayRequest = JsonArrayRequest(url,responseListener,reponseError)
        val queue = Volley.newRequestQueue(this)
        queue.add(jsonArrayRequest)

    }

    companion object{
        lateinit var data: InfoObj
        lateinit var context: Context
    }
}